### ISSUE TITLE
Create singularity container in addition to the docker container

### ISSUE THREAD

**Issue by pkerpedjiev** [higlass/higlass#161]
From @nservant's comment in the cooler repo:

https://github.com/mirnylab/cooler/issues/90#issuecomment-337936279

> Thanks. It make sense.
One more comment, would you plan to generate a singularity container of higlass ?
Docker is great but you need to be root to manage the container. For instance I was very interested in creating a higlass instance open to all teams of my institute but I'm not root on the system ... and singularity offers the possibility to handle container without root rights.
Thanks again for all your help

**Comment by nservant** [comment]
Thanks.
I just build a singularity container for Hic-Pro.
If you want an example to start with, you can have a look at ;
https://github.com/nservant/HiC-Pro/blob/devel/Singularity
And in your case, as you already have a docker container it would be even simpler as I think you can call a docker within a singularity
http://singularity.lbl.gov/docs-docker
Best


Envoyé depuis mon smartphone Samsung Galaxy.


-------- Message d'origine --------
De : Peter Kerpedjiev <notifications@github.com>
Date : 19/10/2017 17:07 (GMT+01:00)
À : hms-dbmi/higlass <higlass@noreply.github.com>
Cc : Servant Nicolas <Nicolas.Servant@curie.fr>, Mention <mention@noreply.github.com>
Objet : [hms-dbmi/higlass] Create singularity container in addition to the docker container (#161)


From @nservant<https://github.com/nservant>'s comment in the cooler repo:

mirnylab/cooler#90 (comment)<https://github.com/mirnylab/cooler/issues/90#issuecomment-337936279>

Thanks. It make sense.
One more comment, would you plan to generate a singularity container of higlass ?
Docker is great but you need to be root to manage the container. For instance I was very interested in creating a higlass instance open to all teams of my institute but I'm not root on the system ... and singularity offers the possibility to handle container without root rights.
Thanks again for all your help

**Comment by pkerpedjiev** [comment]
Thanks! That's very helpful.

**Comment by scriscio** [comment]
We are also interested in using singularity container for deploying higlass.  Please let us know if things move forward.

**Comment by pkerpedjiev** [comment]
May I ask why you're interesting in Singularity over Docker? Is it also the root vs. non-root runtime issue?

**Comment by scriscio** [comment]
We were going to test deployment of higlass on a cluster which supports singularity but currently does not support docker.

**Comment by pkerpedjiev** [comment]
Makes sense. I just tried installing Singularity locally but ran into some issues. I'll try again on another computer at some point in the near future.

In the meantime, if you want you could try building a HiGlass Singularity container yourself. According to the docs (https://singularity.lbl.gov/docs-build-container), you should be able to build one directly from the higlass docker container located here: https://hub.docker.com/r/gehlenborglab/higlass/

If you do end up trying that, I'll be happy to work with you to add the procedure to the documentation.

**Comment by abushoy** [comment]
Hi @pkerpedjiev, we hit an obstacle with higlass docker to singularity conversion.  Please see conversation https://github.com/singularityware/singularity/issues/1939#issuecomment-420107939

**Comment by pkerpedjiev** [comment]
Oh, great! Thanks for the link. If that doesn't get resolved we can try to create a Singularity container from scratch.

**Comment by abushoy** [comment]
@pkerpedjiev  Could you share Dockerfile for the container if possible?

**Comment by pkerpedjiev** [comment]
Yup. It's here: https://github.com/hms-dbmi/higlass-docker/blob/master/web-context/Dockerfile.template

Except the version numbers aren't filled in. To fill them in you need to run `build.sh` from that same repo. If you take out the `docker pull` and `docker build` lines from that file, it will create a complete Dockerfile in `web-context/Dockerfile`.

**Comment by pkerpedjiev** [comment]
@abushoy Did you have any success creating a Singularity container using the Dockerfile?

**Comment by abushoy** [comment]
Hi Peter, 
Unfortunately, docker to singularity conversion didn't yield results. We were able to bypass /bin/sh  error but getting now getting new once :(

**Comment by AlcaArctica** [comment]
This is a very old thread, but I am on the same team. Our system admins have decided that Docker cannot run in the cluster, but we have Singularity. I know that singularity should be compatible with Docker containers. But I just gave it a quick try and could not get it to work. Also fail to see how this would interact with the higlass-manage wrapper...
Anyway, has there been any progress on this? Has any one successfully used higlass inside singularity?

**Comment by rsdmse** [comment]
I tried to help a user here to run this in Apptainer (continuation of Singularity project) but no luck. Following https://github.com/higlass/higlass-docker?tab=readme-ov-file#running-locally I downloaded the cooler file and created the data and tmp directories.

```
# start the instance (equivalent to the 1st code block in readme)
$ apptainer instance start -B /scratch/$USER:/var/log,/scratch/$USER:/var/run,$HOME/hg-data:/data,$HOME/hg-tmp:/tmp ~/higlass-docker_latest.sif hg

# check if the instance is running
$ apptainer instance list
INSTANCE NAME    PID       IP    IMAGE
hg               127248          .../higlass-docker_latest.sif

# check if it sees the cooler file
$ apptainer exec instance://hg ls /tmp
dixon2012-h1hesc-hindiii-allreps-filtered.1000kb.multires.cool

# run the ingest
$ apptainer exec --pwd /home/higlass/projects instance://hg python higlass-server/manage.py ingest_tileset --filename /tmp/$COOLER --filetype cooler --datatype matrix
```
The last command fails with
```
sqlite3.OperationalError: no such table: tilesets_project
```
